server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,915 @@
+			throw new Chat.ErrorMessage(`This user is blocking PMs, and cannot be friended right now.`);
+		}
+
+		const result = await this.query({type: 'transaction', data: [user.id, receiverID], statement: 'send'});
+		if (receiver) {
+			this.sendPM(`/text ${Utils.escapeHTML(user.name)} sent you a friend request!`, receiver.id);
+			this.sendPM(buf, receiver.id);
+			this.sendPM(disclaimer, receiver.id);
+		}
+		this.sendPM(`/text You sent a friend request to ${receiver?.connected ? receiver.name : receiverID}!`, user.id);
+		this.sendPM(
+			`/raw <button class="button" name="send" value="/friends undorequest ${Utils.escapeHTML(receiverID)}">` +
+			`<i class="fa fa-undo"></i> Undo</button>`, user.id
+		);
+		this.sendPM(disclaimer, user.id);
+		return result;
+	}
+	async removeRequest(receiverName: string, senderName: string) {
+		const senderID = toID(senderName);
+		const receiverID = toID(receiverName);
+		if (!senderID) throw new Chat.ErrorMessage(`Invalid sender username.`);
+		if (!receiverID) throw new Chat.ErrorMessage(`Invalid receiver username.`);
+
+		return this.query({
+			statement: 'deleteRequest', data: [senderID, receiverID], type: 'run',
+		});
+	}
+	async approveRequest(receiver: User, senderName: string) {
+		const senderID = toID(senderName);
+		return this.query({type: 'transaction', data: [senderID, receiver.id], statement: 'accept'});
+	}
+	async visualizeList(user: User) {
+		const friends = await this.getFriends(user);
+		if (!friends.size) {
+			return `<h3>Your friends:</h3> <h4>None.</h4>`;
+		}
+		const categorized: {[k: string]: string[]} = {
+			online: [],
+			idle: [],
+			busy: [],
+			offline: [],
+		};
+		for (const friendName of [...friends].sort()) {
+			const friend = Users.get(friendName);
+			if (friend?.connected) {
+				categorized[friend.statusType].push(friend.id);
+			} else {
+				categorized.offline.push(toID(friendName));
+			}
+		}
+
+		const sorted = Object.keys(categorized)
+			.filter(item => categorized[item].length > 0)
+			.map(item => `${STATUS_TITLES[item]} (${categorized[item].length})`);
+
+		let buf = `<h3>Your friends: <small> `;
+		if (sorted.length > 0) {
+			buf += `Total (${friends.size}) | ${sorted.join(' | ')}`;
+		} else {
+			buf += `</h3><em>you have no friends added on Showdown lol</em><br /><br /><br />`;
+			buf += `<strong>To add a friend, use </strong><code>/friend add [username]</code>.<br /><br />`;
+			buf += `<strong>To move over your friends to this account from a different account, `;
+			buf += `sign into that account and use </strong><code>/friend requesttransfer [new name]</code>.`;
+			return buf;
+		}
+		buf += `</h3> `;
+
+		for (const key in friends) {
+			const friendArray = categorized[key].sort();
+			if (friendArray.length === 0) continue;
+			buf += `<h4>${STATUS_TITLES[key]} (${friendArray.length})</h4>`;
+			for (const friend of friendArray) {
+				buf += `<div class="pad"><div>`;
+				buf += await this.displayFriend(friend);
```

Comment:
`displayFriend` calls `getLastLogin`, so this does `O(n)` queries. Surely you already have the last login handy from `getFriends`?